### PR TITLE
feat: Handle Mandatory ModuleTemplates with the new ModuleTemplate naming

### LINKED ISSUE #2068: [Module Catalog] Fetch Mandatory ModuleTemplate which has the highest version
### Description

We agreed to have the name of the mandatory ModuleTemplates the same format as that of the regular ModuleTemplates `modulename-version`. So we need to handle the case when there are multiple ModuleTemplates for the same Mandatory Module by using the ModuleTemplate with the highest version.

### Reasons

_No response_

### Acceptance Criteria

- [x] Add the logic to fetch the ModuleTemplate with highest version for mandatory modules.
- [x] Add an E2E test scenario with two ModuleTemplates with different versions for a mandatory module.  (Validate its installation as well as its deletion).
- [x] Add documentation that for mandatory modules, KLM will install the module template with the highest version in case there are multiple
- [ ] Follow-up on informing neighbors that modulectl can now also be used for mandatory modules

### Feature Testing

Integration tests, End-to-End tests, Unit tests

### Testing approach

_No response_

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by nesmabadr**
**Description**

Changes proposed in this pull request:

- Adjust mandatory module template lookup to fetch the latest version
- Adapt tests

**Related issue(s)**
Resolves #2068

### CODE REVIEW

File: pkg/templatelookup/mandatory.go [hunk 1]
```diff
@@ -25,11 +26,72 @@ func GetMandatory(ctx context.Context, kymaClient client.Reader) (ModuleTemplate
...
+				ModuleTemplate: currentModuleTemplate,
 				Err:            nil,
 			}
 		}
 	}
 	return mandatoryModules, nil
 }
+
+// TODO: Create an issue to remove this function and only use the spec.ModuleName when mandatory modules use modulectl
+func GetModuleName(moduleTemplate *v1beta2.ModuleTemplate) string {
```
Comment (nesmabadr) [inline pkg/templatelookup/mandatory.go:53]: will create issues to remove the fallback after warden ModuleTemplate is created using modulectl

File: tests/e2e/mandatory_module_test.go [hunk 2]
```diff
@@ -137,10 +143,10 @@ var _ = Describe("Mandatory Module Installation and Deletion", Ordered, func() {
 				}).
 				Should(Succeed())
 		})
-		It("Then Kyma is in a \"Error\" State", func() {
+		It("Then Kyma is in a \"Warning\" State", func() {
 			Eventually(KymaIsInState).
 				WithContext(ctx).
-				WithArguments(kyma.GetName(), kyma.GetNamespace(), kcpClient, shared.StateError).
+				WithArguments(kyma.GetName(), kyma.GetNamespace(), kcpClient, shared.StateWarning).
```
Comment (c-pius) [inline tests/e2e/mandatory_module_test.go:149]: Where is this change coming from?

File: .github/actions/deploy-template-operator-with-modulereleasemeta/action.yml [hunk 3]
```diff
@@ -70,14 +79,27 @@ runs:
...
+        ./deploy_moduletemplate.sh ${{ env.ModuleName }} ${{ env.OlderVersionForMandatoryModule }} true true
+
+    - name: Create ModuleTemplate in a new version for Mandatory module
+      if: ${{ matrix.e2e-test == 'mandatory-module'}}
+      working-directory: template-operator
+      shell: bash
+      run: |
+        make build-manifests
+        yq eval '(. | select(.kind == "Deployment") | .metadata.name) = "${{ env.ModuleDeploymentNameInNewerVersion }}"' -i template-operator.yaml
+        ./deploy_moduletemplate.sh ${{ env.ModuleName }} ${{ env.NewerVersionForMandatoryModule }} true true false
```
Comment (c-pius) [inline .github/actions/deploy-template-operator-with-modulereleasemeta/action.yml:101]: Personally I find it a bit cumbersome to lookup the arguments of such scripts. For the create test clsuters one, I also changed it to named arguments: https://github.com/kyma-project/lifecycle-manager/blob/main/scripts/tests/create_test_clusters.

What do you think? Does it make sense for you if we introduce this to all scripts, as a follow-up of course?

File: .github/actions/deploy-lifecycle-manager-e2e/action.yml [hunk 4]
```diff
@@ -109,7 +109,8 @@ runs:
         matrix.e2e-test == 'module-status-decoupling-with-deployment' ||
         matrix.e2e-test == 'purge-metrics' ||
         matrix.e2e-test == 'self-signed-certificate-rotation' ||
-        matrix.e2e-test == 'mandatory-module-metrics'}}  
+        matrix.e2e-test == 'mandatory-module-metrics' ||
+        matrix.e2e-test == 'mandatory-module-metrics-without-version-naming'}}
```
Comment (c-pius) [inline .github/actions/deploy-lifecycle-manager-e2e/action.yml:113]: I was a bit confused about the `without-version-naming` name and had to spend some time to figure out what this is about. Maybe we can rather call it `...with-old-naming-pattern`. Also not super happy with it because the name is actually not important in the processing, but it should be a bit clearer and it will get removed anyway.
Feel free to disagree, we can also keep it as you proposed.

File: docs/contributor/02-controllers.md [hunk 5]
```diff
@@ -40,7 +40,7 @@ Lifecycle Manager uses two Mandatory Modules Controllers:
...
 
-Since the channel concept does not apply to mandatory modules, the Mandatory Modules Installation Controller fetches all the Mandatory ModuleTemplate CRs without any channel filtering. It then translates the ModuleTemplate CR for the mandatory module to a [Manifest CR](../../api/v1beta2/manifest_types.go) with an OwnerReference to the Kyma CR. Similarly to the [Kyma Controller](../../internal/controller/kyma/controller.go),
+Since the channel concept does not apply to mandatory modules, the Mandatory Modules Installation Controller fetches all the Mandatory ModuleTemplate CRs without any channel filtering. The ModuleTemplates with the label 'operator.kyma-project.io/mandatory-module' are fetched, if there are multiple ModuleTemplates for the same mandatory module, the Controller fetches the ModuleTemplate with the highest version. It then translates the ModuleTemplate CR for the mandatory module to a [Manifest CR](../../api/v1beta2/manifest_types.go) with an OwnerReference to the Kyma CR. Similarly to the [Kyma Controller](../../internal/controller/kyma/controller.go),
```
[the remark this answers is not in the document]
Comment (nesmabadr) [reply docs/contributor/02-controllers.md:43]: hmm do you have suggestions how? I am unable to rephrase it to something simpler

File: pkg/templatelookup/mandatory.go [hunk 6]
```diff
@@ -25,11 +26,71 @@ func GetMandatory(ctx context.Context, kymaClient client.Reader) (ModuleTemplate
...
+				moduleTemplate.Spec.Version, err)
+		}
+		return version, nil
+	}
+
+	version, err := semver.NewVersion(moduleTemplate.Annotations[shared.ModuleVersionAnnotation])
+	if err != nil {
+		return nil, fmt.Errorf("could not parse version as a semver %s: %w",
+			moduleTemplate.Annotations[shared.ModuleVersionAnnotation], err)
+	}
```
[the remark this answers is not in the document]
Comment (nesmabadr) [reply pkg/templatelookup/mandatory.go:74]: hmm should we add a TODO or are just the followup tickets enough?

File: pkg/templatelookup/mandatory_test.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+	"github.com/kyma-project/lifecycle-manager/api/v1beta2"
+	"github.com/kyma-project/lifecycle-manager/pkg/templatelookup"
+	"github.com/kyma-project/lifecycle-manager/pkg/testutils/builder"
+)
+
+func TestGetDesiredModuleTemplateForMultipleVersions_ReturnCorrectValue(t *testing.T) {
+	firstModuleTemplate := builder.NewModuleTemplateBuilder().
+		WithName("warden-1.0.0-dev").
+		WithVersion("1.0.0-dev").
+		WithLabel("module-diff", "first").
```
[the remark this answers is not in the document]
Comment (nesmabadr) [reply pkg/templatelookup/mandatory_test.go:20]: I was just adding some different metadata to both ModuleTemplates

File: docs/contributor/02-controllers.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (c-pius) [reply docs/contributor/02-controllers.md:43]: The suggestion is already there :)

File: pkg/templatelookup/mandatory_test.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+		Build()
+
+	secondModuleTemplate := builder.NewModuleTemplateBuilder().
+		WithName("warden-1.0.1-dev").
+		WithVersion("1.0.1-dev").
+		WithLabel("module-diff", "second").
+		Build()
```
[the remark this answers is not in the document]
Comment (nesmabadr) [reply pkg/templatelookup/mandatory_test.go:27]: I am not sure I get you mean, can you explain please?

File: tests/e2e/mandatory_module_test.go (hunk 2 above)
Comment (nesmabadr) [reply tests/e2e/mandatory_module_test.go:149]: because the new ModuleTemplates are not bound to channel, so if you try to enable a mandatory module and it has to be bound to channel in the skr cluster it will just show a warning that there are no ModuleTemplates for this channel

File: pkg/templatelookup/mandatory.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (c-pius) [reply pkg/templatelookup/mandatory.go:74]: I usually do both actually: https://github.com/kyma-project/lifecycle-manager/blob/main/internal/remote/remote_catalog.go#L90-L96

Then it is clear from the code and we can also use a search to find all known places to touch from the ticket.

File: .github/actions/deploy-template-operator-with-modulereleasemeta/action.yml (hunk 3 above)
Comment (nesmabadr) [reply .github/actions/deploy-template-operator-with-modulereleasemeta/action.yml:101]: yes will do so, I wasn't sure how to have them named to be honest :D

File: .github/actions/deploy-lifecycle-manager-e2e/action.yml (hunk 4 above)
Comment (nesmabadr) [reply .github/actions/deploy-lifecycle-manager-e2e/action.yml:113]: yes I was hesitanta lot about the name and even thought about old_naming_.. but then I thought I would just write the difference. but since you have the same thought I will rename it

File: pkg/templatelookup/mandatory_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (c-pius) [reply pkg/templatelookup/mandatory_test.go:27]: In the downstream function where we get the module semver we have one branch looking for spec.version, the other for module-version annotatoin/label. My idea was that we use both apporaches in this test to have both branches covered.

File: .github/actions/deploy-template-operator-with-modulereleasemeta/action.yml (hunk 3 above)
Comment (c-pius) [reply .github/actions/deploy-template-operator-with-modulereleasemeta/action.yml:101]: Also took some help from copilot for this :D

File: tests/e2e/mandatory_module_test.go (hunk 2 above)
Comment (c-pius) [reply tests/e2e/mandatory_module_test.go:149]: Understood. Just wondering if we really want Warning for this case. But could be follow-up upon in a separate issue.

File: docs/contributor/02-controllers.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (nesmabadr) [reply docs/contributor/02-controllers.md]: ahh sorry my fault I only saw the If change :sweat_smile:

File: .github/actions/deploy-template-operator-with-modulereleasemeta/action.yml (hunk 3 above)
Comment (nesmabadr) [reply .github/actions/deploy-template-operator-with-modulereleasemeta/action.yml:101]: https://github.com/kyma-project/lifecycle-manager/issues/2136

**Review by c-pius — DISMISSED** [review]

File: tests/e2e/mandatory_module_test.go (hunk 2 above)
Comment (c-pius) [reply tests/e2e/mandatory_module_test.go:149]: FYI: https://github.com/kyma-project/lifecycle-manager/issues/2137

File: docs/contributor/02-controllers.md [hunk 9]
```diff
@@ -40,7 +40,7 @@ Lifecycle Manager uses two Mandatory Modules Controllers:
...
 * [Mandatory modules deletion controller](../../internal/controller/mandatorymodule/deletion_controller.go) deals with the deletion of mandatory modules
 
-Since the channel concept does not apply to mandatory modules, the Mandatory Modules Installation Controller fetches all the Mandatory ModuleTemplate CRs without any channel filtering. It then translates the ModuleTemplate CR for the mandatory module to a [Manifest CR](../../api/v1beta2/manifest_types.go) with an OwnerReference to the Kyma CR. Similarly to the [Kyma Controller](../../internal/controller/kyma/controller.go),
+Since the channel concept does not apply to mandatory modules, the Mandatory Modules Installation Controller fetches all the Mandatory ModuleTemplate CRs with the label 'operator.kyma-project.io/mandatory-module'. If there are multiple ModuleTemplates for the same mandatory module, the Controller fetches the ModuleTemplate with the highest version. It then translates the ModuleTemplate CR for the mandatory module to a [Manifest CR](../../api/v1beta2/manifest_types.go) with an OwnerReference to the Kyma CR. Similarly to the [Kyma Controller](../../internal/controller/kyma/controller.go),
```
Comment (mmitoraj) [inline docs/contributor/02-controllers.md:43]:
```suggestion
Since the channel concept does not apply to mandatory modules, the Mandatory Modules Installation Controller fetches all the Mandatory ModuleTemplate CRs with the 'operator.kyma-project.io/mandatory-module' label. If multiple ModuleTemplates exist for the same mandatory module, the Controller fetches the ModuleTemplate with the highest version. It then translates the ModuleTemplate CR for the mandatory module to a [Manifest CR](../../api/v1beta2/manifest_types.go) with an OwnerReference to the Kyma CR. Similarly to the [Kyma Controller](../../internal/controller/kyma/controller.go),
```

**Review by mmitoraj — APPROVED** [review]

**Review by c-pius — APPROVED** [review]
